src/server/frontend_wayland/text_input_v1.cpp

```diff
@@ -0,0 +1,421 @@
+    return std::make_shared<TextInputManagerV1Global>(display, std::move(ctx));
+}
+
+TextInputManagerV1Global::TextInputManagerV1Global(
+    wl_display* display,
+    std::shared_ptr<TextInputV1Ctx> ctx)
+    : Global{display, Version<1>()},
+      ctx{std::move(ctx)}
+{
+}
+
+void TextInputManagerV1Global::bind(wl_resource* new_resource)
+{
+    new TextInputManagerV1{new_resource, ctx};
+}
+
+TextInputManagerV1::TextInputManagerV1(
+    wl_resource *resource,
+    std::shared_ptr<TextInputV1Ctx> ctx)
+    : mw::TextInputManagerV1{resource, Version<1>()},
+      ctx{std::move(ctx)}
+{
+}
+
+void TextInputManagerV1::create_text_input(wl_resource *id)
+{
+    new TextInputV1{id, ctx};
+}
+
+TextInputV1::TextInputV1(
+    wl_resource* resource,
+    std::shared_ptr<TextInputV1Ctx> const& ctx)
+    : mw::TextInputV1{resource, Version<1>()},
+      ctx{ctx},
+      handler{std::make_shared<Handler>(this, ctx->wayland_executor)}
+{
+}
+
+TextInputV1::~TextInputV1()
+{
+    if (seat)
+    {
+        seat.value()->remove_focus_listener(client, this);
+        delete seat.value();
+    }
+
+    on_new_input_field = false;
+    pending_state.reset();
+}
+
+void TextInputV1::send_text_change(ms::TextInputChange const& change)
+{
+    auto const client_serial = find_client_serial(change.serial);
+    if (!pending_state || !current_surface || !client_serial)
+    {
+        // We are no longer enabled, or we don't have a valid serial
+        return;
+    }
+    if (change.preedit_text || change.preedit_cursor_begin || change.preedit_cursor_end)
+    {
+        send_preedit_cursor_event(change.preedit_cursor_begin.value_or(0));
+        send_preedit_string_event(client_serial.value(), change.preedit_text.value_or(""), "");
+    }
+    if (change.delete_before || change.delete_after)
+    {
+        send_delete_surrounding_text_event(change.delete_before.value_or(0), change.delete_after.value_or(0));
+    }
+    if (change.commit_text)
+    {
+        send_commit_string_event(client_serial.value(), change.commit_text.value());
+    }
+}
+
+auto TextInputV1::find_client_serial(ms::TextInputStateSerial state_serial) const -> std::optional<uint32_t>
+{
+    // Loop in reverse order because the serial we're looking for will generally be at the end
+    for (auto it = state_serials.rbegin(); it != state_serials.rend(); it++)
+    {
+        if (it->second == state_serial)
+        {
+            return it->first;
+        }
+    }
+    return std::nullopt;
+}
+
+void TextInputV1::focus_on(mf::WlSurface* surface)
+{
+    if (current_surface)
+    {
+        send_leave_event();
+    }
+    current_surface = mw::make_weak(surface);
+    if (surface)
+    {
+        send_enter_event(surface->resource);
+    }
+    else
+    {
+        deactivate(nullptr); // TODO - Is this right?
+    }
+}
+
+void TextInputV1::activate(wl_resource *seat, wl_resource *surface)
```

Comment:
I'd call `seat` `seat_resource` or something (or change the name of `TextInputV1::seat` to avoid the explicit `this` below,